### PR TITLE
SAT Progress

### PR DESCRIPTION
**Pull request by jmgomez**
Most tests are green now

### CODE REVIEW

File: src/nimble.nim [hunk 1]
```diff
@@ -72,47 +72,65 @@ proc addReverseDeps(solvedPkgs: seq[SolvedPackage], allPkgsInfo: seq[PackageInfo
...
   if solvedPkgs.len > 0: 
     displaySatisfiedMsg(solvedPkgs, pkgsToInstall)
     addReverseDeps(solvedPkgs, allPkgsInfo, options)
-    return result
+    for pkg in allPkgsInfo:
+      result.incl pkg
+    result = 
+      result.toSeq
+      .deleteStaleDependencies(rootPkgInfo, options)
+      .deduplicate.toHashSet
```
Comment (Araq) [inline src/nimble.nim:101]: Quite sure that `toHashSet` does the deduplicate step, and more efficiently too.

File: src/nimble.nim [hunk 2]
```diff
@@ -72,47 +72,65 @@ proc addReverseDeps(solvedPkgs: seq[SolvedPackage], allPkgsInfo: seq[PackageInfo
...
+    satProccesedPackages = result
+    return
```
Comment (Araq) [inline src/nimble.nim:103]:
```suggestion
    return result
```

For clarity.

File: src/nimble.nim [hunk 3]
```diff
@@ -72,47 +72,65 @@ proc addReverseDeps(solvedPkgs: seq[SolvedPackage], allPkgsInfo: seq[PackageInfo
...
   addReverseDeps(solvedPkgs, allPkgsInfo, options)
 
+  result = deleteStaleDependencies(result.toSeq, rootPkgInfo, options).deduplicate.toHashSet  
+  satProccesedPackages = result
+
   if not solved:
     display("Error", output, Error, priority = HighPriority)
     raise nimbleError("Unsatisfiable dependencies")
-  
+
```
[the remark this answers is not in the document]
Comment (Araq) [reply src/nimble.nim:133]: No trailing whitespace please.

File: src/nimble.nim [hunk 4]
```diff
@@ -477,19 +493,20 @@ proc installFromDir(dir: string, requestedVer: VersionRange, options: Options,
...
+    if not options.useSatSolver: #The dep path is not created when using the sat solver as packages are collected upfront
+      var oldPkg = oldPkg.get
+      oldPkg.metaData.specialVersions.incl pkgInfo.metaData.specialVersions
+      saveMetaData(oldPkg.metaData, oldPkg.getNimbleFileDir, changeRoots = false)
+      if result.deps.contains oldPkg:
+        result.deps[oldPkg].metaData.specialVersions.incl(
+          oldPkg.metaData.specialVersions)
+      result.deps.incl oldPkg
+      result.pkg = oldPkg
+      return
```
Comment (Araq) [inline src/nimble.nim:509]:
```suggestion
      return result
```

File: src/nimblepkg/nimblesat.nim [hunk 5]
```diff
@@ -323,7 +323,8 @@ proc getSolvedPackages*(pkgVersionTable: Table[string, PackageVersions], output:
         result.add solvedPkg
 
 proc getCacheDownloadDir*(url: string, ver: VersionRange, options: Options): string =
-  options.pkgCachePath / getDownloadDirName(url, ver, notSetSha1Hash)
+  return ""
+  # options.pkgCachePath / getDownloadDirName(url, ver, notSetSha1Hash)
```
Comment (Araq) [inline src/nimblepkg/nimblesat.nim:327]: Why was this disabled?

Comment (jmgomez) [reply src/nimblepkg/nimblesat.nim:327]: It prevents packages being downloaded twice (`""` makes it so it uses a temp directory). A test is failing when is set (SAT only oc). I disabled it so I can be sure it doesnt interfere with the main goal which is making existing functionality work with SAT, after that I will work on the failing test (spend a lot of time already trying to fix it, but decided to move on and fix SAT first)

**Pull request merged**